File: Sources/Switcher3w/PhraseTracker.swift
import Foundation

/// Phrase-level language memory for auto-conversion (phrase-aware-ambiguity).
///
/// A "phrase" is the run of evaluated words since the last hard reset (Enter/Tab/arrows/
/// mouse click/app or focus switch — the same events that reset the word buffer).
/// Words whose uk/ru ambiguity was resolved by the preference are remembered as `defaulted`;
/// when a later word is valid in exactly ONE language, the defaulted words of other languages
/// are re-converted to it in a single segment replacement. Precision-first: the tracker
/// resets on anything it can't account for exactly, and a phrase locked to one language
/// never corrects toward another.
@MainActor
final class PhraseTracker {
    enum WordKind: Equatable {
        case defaulted(lang: String)  // ambiguity resolved by preference/lock — retro-correctable
        case locked(lang: String)     // valid in exactly one language — locks the phrase
        case neutral                  // kept / valid as typed — reproduced verbatim in corrections
    }

    struct PhraseWord {
        let keys: [TypedKey]
        var shownText: String   // what this word looks like on screen right now
        var spacesAfter: Int
        var kind: WordKind
    }

    /// Maximum characters a correction may erase — bounds worst-case erase chains.
    static let maxCorrectionLength = 200

    private(set) var words: [PhraseWord] = []
    /// Bumped on every reset. Retype completions run asynchronously; a completion whose
    /// captured generation no longer matches means a reset (click/Enter/focus change) raced
    /// it — the screen has moved on and the pending record/confirm must be dropped.
    private(set) var generation = 0

    /// The language the phrase is locked to (first exactly-one-language word), nil if none yet.
    var lockedLang: String? {
        for w in words { if case .locked(let lang) = w.kind { return lang } }
        return nil
    }

    func reset() {
        if !words.isEmpty { rslog("phrase: reset (\(words.count) word(s))") }
        words = []
        generation &+= 1
    }

    /// Records an evaluated word. Pass `ifGeneration` from an async completion so a record
    /// that lost the race against a reset is dropped instead of corrupting the phrase.
    func record(keys: [TypedKey], shownText: String, spacesAfter: Int,
                kind: WordKind, ifGeneration gen: Int? = nil) {
        if let gen, gen != generation {
            rslog("phrase: record dropped (stale generation)")
            return
        }
        words.append(PhraseWord(keys: keys, shownText: shownText, spacesAfter: spacesAfter, kind: kind))
    }

    /// An extra space arrived after the last recorded word (no new word between) — keeps the
    /// segment character math exact for multi-space runs.
    func noteExtraSpace() {
        guard !words.isEmpty else { return }
        words[words.count - 1].spacesAfter += 1
    }

    /// A planned retro-correction: the on-screen segment to erase, its replacement, and the
    /// updated word records to store once the replacement succeeded.
    struct Correction {
        let oldSegment: String
        let newSegment: String
        let firstIndex: Int
        let correctedWords: [PhraseWord]
    }

    /// Builds the correction toward `lang` (rendered through the layout `layoutID`): the
    /// segment from the first defaulted-to-another-language word through the last recorded
    /// word. Defaulted words re-render their keystrokes; neutral/locked words are reproduced
    /// verbatim. nil — nothing defaulted to another language, the phrase is locked to a
    /// conflicting language, a re-render failed, or the segment exceeds the length cap.
    func correction(toLang lang: String, layoutID: String) -> Correction? {
        if let locked = lockedLang, locked != lang {
            rslog("phrase: contradictory (locked \(locked), new \(lang)) — no correction")
            return nil
        }
        guard let first = words.firstIndex(where: {
            if case .defaulted(let l) = $0.kind { return l != lang }
            return false
        }) else { return nil }

        var old = "", new = ""
        var corrected: [PhraseWord] = []
        for index in first..<words.count {
            var word = words[index]
            let spaces = String(repeating: " ", count: word.spacesAfter)
            old += word.shownText + spaces
            if case .defaulted(let l) = word.kind, l != lang {
                guard let rerendered = NWayResolver.render(keys: word.keys, layoutID: layoutID) else {
                    rslog("phrase: re-render failed — no correction")
                    return nil
                }
                word.shownText = rerendered
                word.kind = .defaulted(lang: lang)
            }
            new += word.shownText + spaces
            corrected.append(word)
        }
        guard old.count <= Self.maxCorrectionLength else {
            rslog("phrase: correction segment too long (\(old.count) chars) — skipped")
            return nil
        }
        return Correction(oldSegment: old, newSegment: new, firstIndex: first, correctedWords: corrected)
    }

    /// Commits a successful correction to the memory (call from the retype success completion,
    /// with the generation captured when the correction was planned).
    func confirm(_ correction: Correction, ifGeneration gen: Int) {
        guard gen == generation else {
            rslog("phrase: confirm dropped (stale generation)")
            return
        }
        guard correction.firstIndex + correction.correctedWords.count == words.count else {
            reset()   // the phrase changed shape while the retype ran — memory is unreliable
            return
        }
        words.replaceSubrange(correction.firstIndex..., with: correction.correctedWords)
        rslog("phrase: correction committed (\(correction.correctedWords.count) word(s))")
    }
}

File: windows/src/Switcher3way.Core/PhraseTracker.cs
namespace Switcher3way.Core;

/// <summary>
/// Phrase-level language memory for auto-conversion (phrase-aware ambiguity). A faithful port of the
/// macOS <c>PhraseTracker</c>.
///
/// A "phrase" is the run of evaluated words since the last hard reset (Enter/Tab/arrows/mouse click/
/// app or focus switch — the same events that reset the word buffer). Words whose uk/ru ambiguity was
/// resolved by the preference are remembered as <see cref="WordKind.Defaulted"/>; when a later word is
/// valid in exactly one language, the defaulted words of other languages are re-converted to it in a
/// single segment replacement. Precision-first: the tracker resets on anything it can't account for
/// exactly, and a phrase locked to one language never corrects toward another.
///
/// Not thread-safe by design — the Windows engine mutates it only on its single worker thread
/// (word events and resets are both marshaled there), so no locking is needed.
/// </summary>
public sealed class PhraseTracker
{
    /// <summary>Renders a word's keys through a layout id (backed by <c>NWayResolver.Render</c>).</summary>
    public delegate string? RenderFunc(IReadOnlyList<TypedKey> keys, string layoutId);

    private readonly RenderFunc _render;
    public PhraseTracker(RenderFunc render) => _render = render;

    public abstract record WordKind
    {
        private WordKind() { }
        /// <summary>Ambiguity resolved by preference/lock — retro-correctable.</summary>
        public sealed record Defaulted(string Lang) : WordKind;
        /// <summary>Valid in exactly one language — locks the phrase.</summary>
        public sealed record Locked(string Lang) : WordKind;
        /// <summary>Kept / valid as typed — reproduced verbatim in corrections.</summary>
        public sealed record Neutral : WordKind;
    }

    public sealed class PhraseWord
    {
        public required IReadOnlyList<TypedKey> Keys { get; init; }
        public required string ShownText { get; set; } // what this word looks like on screen right now
        public required int SpacesAfter { get; set; }
        public required WordKind Kind { get; set; }
    }

    /// <summary>Maximum characters a correction may erase — bounds worst-case erase chains.</summary>
    public const int MaxCorrectionLength = 200;

    private readonly List<PhraseWord> _words = new();
    public IReadOnlyList<PhraseWord> Words => _words;

    /// <summary>Bumped on every reset; a stale generation drops a record/confirm that lost a race.</summary>
    public int Generation { get; private set; }

    /// <summary>The language the phrase is locked to (first exactly-one-language word), null if none.</summary>
    public string? LockedLang =>
        _words.Select(w => w.Kind).OfType<WordKind.Locked>().Select(l => l.Lang).FirstOrDefault();

    public void Reset()
    {
        _words.Clear();
        Generation++;
    }

    /// <summary>Records an evaluated word. Pass <paramref name="ifGeneration"/> from a deferred caller
    /// so a record that lost the race against a reset is dropped instead of corrupting the phrase.</summary>
    public void Record(IReadOnlyList<TypedKey> keys, string shownText, int spacesAfter,
                       WordKind kind, int? ifGeneration = null)
    {
        if (ifGeneration is int g && g != Generation) return;
        _words.Add(new PhraseWord { Keys = keys, ShownText = shownText, SpacesAfter = spacesAfter, Kind = kind });
    }

    /// <summary>An extra space arrived after the last word (no new word between) — keeps the segment
    /// character math exact for multi-space runs.</summary>
    public void NoteExtraSpace()
    {
        if (_words.Count > 0) _words[^1].SpacesAfter++;
    }

    /// <summary>A planned retro-correction: the on-screen segment to erase, its replacement, the index
    /// it starts at, and the updated word records to store once the replacement succeeded.</summary>
    public sealed record Correction(string OldSegment, string NewSegment, int FirstIndex,
                                    IReadOnlyList<PhraseWord> CorrectedWords);

    /// <summary>
    /// Builds the correction toward <paramref name="lang"/> (rendered through <paramref name="layoutId"/>):
    /// the segment from the first word defaulted to a *different* language through the last recorded
    /// word. Defaulted words re-render their keystrokes; neutral/locked words are reproduced verbatim.
    /// Null when nothing is defaulted to another language, the phrase is locked to a conflicting
    /// language, a re-render fails, or the segment exceeds the length cap.
    /// </summary>
    public Correction? BuildCorrection(string lang, string layoutId)
    {
        if (LockedLang is string locked && locked != lang) return null;

        var first = -1;
        for (var i = 0; i < _words.Count; i++)
            if (_words[i].Kind is WordKind.Defaulted d && d.Lang != lang) { first = i; break; }
        if (first < 0) return null;

        var old = ""; var neo = "";
        var corrected = new List<PhraseWord>();
        for (var index = first; index < _words.Count; index++)
        {
            var w = _words[index];
            var spaces = new string(' ', w.SpacesAfter);
            old += w.ShownText + spaces;

            var shown = w.ShownText;
            var kind = w.Kind;
            if (w.Kind is WordKind.Defaulted dd && dd.Lang != lang)
            {
                var rerendered = _render(w.Keys, layoutId);
                if (rerendered is null) return null;
                shown = rerendered;
                kind = new WordKind.Defaulted(lang);
            }
            neo += shown + spaces;
            corrected.Add(new PhraseWord { Keys = w.Keys, ShownText = shown, SpacesAfter = w.SpacesAfter, Kind = kind });
        }
        if (old.Length > MaxCorrectionLength) return null;
        return new Correction(old, neo, first, corrected);
    }

    /// <summary>Commits a successful correction to the memory (with the generation captured when the
    /// correction was planned).</summary>
    public void Confirm(Correction correction, int ifGeneration)
    {
        if (ifGeneration != Generation) return;
        if (correction.FirstIndex + correction.CorrectedWords.Count != _words.Count)
        {
            Reset(); // the phrase changed shape while the retype ran — memory is unreliable
            return;
        }
        _words.RemoveRange(correction.FirstIndex, _words.Count - correction.FirstIndex);
        _words.AddRange(correction.CorrectedWords);
    }
}
